Editor/Gui/UiSymbolJson.cs

```diff
@@ -19,216 +19,192 @@
 
 namespace T3.Editor.Gui
 {
-    public class SymbolUiJson 
+    public static class SymbolUiJson
     {
-        public JsonTextWriter Writer { get; set; }
-        public JsonTextReader Reader { get; init; }
-        
-        public void WriteSymbolUi(SymbolUi symbolUi)
+        public static void WriteSymbolUi(SymbolUi symbolUi, JsonTextWriter writer)
         {
-            Writer.WriteStartObject();
+            writer.WriteStartObject();
 
-            Writer.WriteObject("Id", symbolUi.Symbol.Id);
-            Writer.WriteComment(symbolUi.Symbol.Name);
+            writer.WriteObject(JsonKeys.Id, symbolUi.Symbol.Id);
+            writer.WriteComment(symbolUi.Symbol.Name);
 
-            Writer.WriteObject("Description", symbolUi.Description);
+            writer.WriteObject(JsonKeys.Description, symbolUi.Description);
 
-            WriteInputUis(symbolUi);
-            WriteChildUis(symbolUi);
-            WriteOutputUis(symbolUi);
-            WriteAnnotations(symbolUi);
+            WriteInputUis(symbolUi, writer);
+            WriteChildUis(symbolUi, writer);
+            WriteOutputUis(symbolUi, writer);
+            WriteAnnotations(symbolUi, writer);
 
-            Writer.WriteEndObject();
+            writer.WriteEndObject();
         }
 
-        private void WriteInputUis(SymbolUi symbolUi)
+        private static void WriteInputUis(SymbolUi symbolUi, JsonTextWriter writer)
         {
-            Writer.WritePropertyName("InputUis");
-            Writer.WriteStartArray();
+            writer.WritePropertyName(JsonKeys.InputUis);
+            writer.WriteStartArray();
 
             foreach (var inputEntry in symbolUi.InputUis)
             {
-                var symbolInput = symbolUi.Symbol.InputDefinitions.SingleOrDefault(inputDef => inputDef.Id == inputEntry.Key);
+                var symbolInput = symbolUi.Symbol.InputDefinitions.FirstOrDefault(inputDef => inputDef.Id == inputEntry.Key);

```

Comment:
I agree in the principle of failing often with clear errors, but I personally would prefer this to happen on the root-most level, such that the error propogates so you don't have to go searching for its origin and consuming classes don't need to constantly check for valid data.

With json, this stance becomes tricker, since anyone can just open the file and ruin it. then it becomes more of a moral question... 😈

Reply:
replaced instances of `FirstOrDefault` with `SingleOrDefault`